### ISSUE TITLE
Add Module-conforming multi-head self-attention layer (MultiHeadAttention)

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#5637]
## Summary

Add a `Module`-conforming **multi-head scaled dot-product self-attention** layer (`MultiHeadAttention`) under `src/odyssey/core/layers/`, alongside the existing `FeedForward`, `GRUCell`, `LayerNorm`, etc. This is the stateful *layer* counterpart to the existing stateless functional attention in `src/odyssey/core/attention.mojo` — mirroring the repo's `activation.mojo` (functional) vs `relu.mojo` (layer) split.

## Motivation

The Transformer feed-forward sub-layer (`FeedForward`) already lands as a first-class layer so a Transformer block can hold it as a sub-`Module` and optimizers can collect its parameters uniformly. The attention sub-layer is the missing half. A standalone, parameter-owning attention block enables composing a full Transformer encoder block from `Module` primitives (`MultiHeadAttention` + `LayerNorm` + `FeedForward`) without reaching into the stateless functional API and hand-managing weights via `MultiHeadAttentionWeights`.

## Proposed API

```mojo
struct MultiHeadAttention[dtype: DType = DType.float32](Copyable, Module, Movable):
    def __init__(out self, d_model: Int, num_heads: Int = 1, causal: Bool = False) raises
    def forward(mut self, input: AnyTensor) raises -> AnyTensor   # [batch, seq, d_model] -> same
    def parameters(self) raises -> List[AnyTensor]                # 8: q/k/v/out weight+bias
    def train(mut self); def eval(mut self)
```

- **Self-attention**, batch-first `[batch, seq, d_model]` tensors.
- Four `Linear(d_model, d_model)` projections (query, key, value, output).
- `Attention(Q, K, V) = softmax(Q Kᵀ / √d_k) V` (Vaswani et al. 2017, arXiv:1706.03762 §3.2); `num_heads` splits `d_model` into heads of width `d_k = d_model / num_heads` (§3.2.2); `num_heads = 1` (default) = single-head.
- Optional causal (autoregressive) `−∞` upper-triangular mask (§3.2.3).
- No residual / no LayerNorm — the bare primitive, composed externally.
- No `BatchNorm`; float32 API throughout.

## Testing

`tests/odyssey/core/layers/test_attention.mojo` (+ `parity_refs/attention_parity_reference.py` generator and committed JSON fixture), following the `FeedForward`/`GRU` parity convention: ramp-seeded weights, torch reference to 1e-5, two cases (single-head non-causal, multi-head causal). Registered in `scripts/run_primitive_test.sh` (`attention`) and the README primitive table.

## Reference

Vaswani, A., et al. (2017). *Attention Is All You Need.* NeurIPS 2017. arXiv:1706.03762, §3.2.
